### PR TITLE
Origin/dev silencemedia factory

### PR DESCRIPTION
**Pull request by Clement-Jean**
## What

Added a Factory to SilenceMediaSource to be able to create it like the other MediaSource. Also added a tag in the class so that users can get more information on this object.

## Why

The idea comes from the following [issue](https://github.com/google/ExoPlayer/issues/7231). In the use case, the user would like to create a transition between two medias in a playlist. To do that we could use a SilenceMediaSource with a tag in order to know more about the media in the `onPositionDiscontinuity()` method. The user would only have to call `player.getCurrentTag()`.

### CODE REVIEW

**Comment by Clement-Jean** [comment]
@googlebot I signed it!

Sent from Mail<https://go.microsoft.com/fwlink/?LinkId=550986> for Windows 10

________________________________

**Comment by icbaker** [comment]
The diff in this PR looks enormous (426 files & 200 commits) - it seems you might not have merged/rebased correctly and a lot of irrelevant changes (by exoplayer-team) are showing up.

Can you fix it so that the diff only shows the files you intend to change?

**Comment by Clement-Jean** [comment]
@icbaker sorry, I apparently did a mistake. Do you have any clue how to change this in this pull request ?

**Review by marcbaechinger — COMMENTED** [review]
In `prepareSourceInternal(@Nullable TransferListener mediaTransferListener)` you need to pass the `tag` to the constructor of the `SinglePeriodTimeline` as the last parameter. This makes the tag available in the timeline.

**Comment by Clement-Jean** [comment]
How people are supposed to know that ?

**Review by marcbaechinger — COMMENTED** [review]
Developers who implement the `MediaSource` interface know they have to provide a timeline with the information of the media source. The java doc of the `MediaSource` interface tells them to provide the timeline.

File: library/core/src/main/java/com/google/android/exoplayer2/source/SilenceMediaSource.java [hunk 1]
```diff
@@ -33,6 +33,30 @@
 /** Media source with a single period consisting of silent raw audio of a given duration. */
 public final class SilenceMediaSource extends BaseMediaSource {
 
+  public static final class Factory {
```
Comment (marcbaechinger) [inline library/core/src/main/java/com/google/android/exoplayer2/source/SilenceMediaSource.java:37]: Please provide a short java doc for public constructors and methods which are not overridden. Here and below.

File: library/core/src/main/java/com/google/android/exoplayer2/source/SilenceMediaSource.java [hunk 2]
```diff
@@ -33,6 +33,30 @@
...
+
+    @Nullable private Object tag;
+    private long duration;
+    private boolean isCreateCalled;
+
+    public SilenceMediaSource.Factory setTag(Object tag) {
+      Assertions.checkState(!isCreateCalled);
```
Comment (marcbaechinger) [inline library/core/src/main/java/com/google/android/exoplayer2/source/SilenceMediaSource.java:52]: We made factories of other media sources re-usable. So create can be called multiple times.  I'd probably remove this for consistency.

Comment (Clement-Jean) [reply library/core/src/main/java/com/google/android/exoplayer2/source/SilenceMediaSource.java:52]: I just followed how the ProgressiveMediaSource.Factory is implemented. So should I remove the boolean and the assertion ?

**Comment by icbaker** [comment]
> @icbaker sorry, I apparently did a mistake. Do you have any clue how to change this in this pull request ?

I think you might have started from the `release-v2` branch but proposed this PR back to the `dev-v2` branch. We do all our development on the `dev-v2` branch, but cherrypick across to `release-v2` to support the point releases - all the spurious commits in this PR seem to be from `release-v2`.

You could try rebasing your current branch onto the head of `dev-v2`. It might work, or it might be a disaster.

If it's a disaster, I'd suggest starting again from `dev-v2`, creating a new branch, and then cherrypicking your changes across from this branch. You'll probably have to then submit a new PR, but that's fine we can just abandon this one.

**Comment by Clement-Jean** [comment]
@icbaker  I can just close this PR and restart right ? I'm just trying to finish my work with @marcbaechinger

**Comment by icbaker** [comment]
> @icbaker I can just close this PR and restart right ? I'm just trying to finish my work with @marcbaechinger

Yeah, that should be totally fine. @marcbaechinger will understand - if you address his comments so far and raise a new PR, we'll be able to follow along. Thanks!

**Pull request closed without merging**